text/0000-consensus-api.md

```diff
@@ -0,0 +1,320 @@
+- Feature Name: consensus-api
+- Start Date: 2018-03-15
+- RFC PR: (leave this empty)
+- Sawtooth Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC describes a new Consensus API for supporting both the existing
+_Nakamoto-style_ consensus algorithms as _elected leader_ algorithms such as
+rBFT.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for this RFC is to support integrating new non-lottery-based
+consensus mechanisms with Sawtooth, such as rBFT, while continuing to support
+PoET/SGX, PoET/Simulator, and devmode consensus modules.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Consensus API was developed after researching a variety of consensus
+algorithms, consensus interfaces, and the consensus problem in general. Out of
+this research came the following observations:
+
+1. All consensus algorithms describe a state machine.
+2. All consensus algorithm state transitions occur as a result of:
+
+    a. Receiving a consensus message from a peer
+    b. Receiving a message from a client
+    c. An internal interrupt
+
+These observations led to the creation of the _Consensus Engine_ abstraction,
+which can be used to implement any consensus algorithm that is either an
+_elected leader_ algorithm or a _Nakamoto-style_ algorithm.
```

Comment:
The hyperledger architecture paper on consensus breaks these methods into _lottery_ (PoET, POW, etc) and _voting_ (PBFT, RBFT, RAFT, etc) based elections. The use of _elected leader_ could be misunderstood.